src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingTaggerProvider.RenameTrackingCodeAction.cs

```diff
@@ -20,57 +20,62 @@ private class RenameTrackingCodeAction : CodeAction
+            }
+
+            protected override async Task<IEnumerable<CodeActionOperation>> ComputePreviewOperationsAsync(CancellationToken cancellationToken)
+            {
+                InitializeRenameTrackingCommitter(cancellationToken);
+                var solutionSet = await _renameTrackingCommitter.RenameSymbolAsync(cancellationToken).ConfigureAwait(false);
+                return SpecializedCollections.SingletonEnumerable((CodeActionOperation)new ApplyChangesOperation(solutionSet.RenamedSolution));
+            }
+
+            private void InitializeRenameTrackingCommitter(CancellationToken cancellationToken)
+            {
                 SourceText text;
                 StateMachine stateMachine;
-                ITextBuffer textBuffer;
 
                 if (_document.TryGetText(out text))
                 {
-                    textBuffer = text.Container.TryGetTextBuffer();
-                    if (textBuffer == null)
-                    {
-                        Environment.FailFast(string.Format("document with name {0} is open but textBuffer is null. Textcontainer is of type {1}. SourceText is: {2}",
-                                                            _document.Name, text.Container.GetType().FullName, text.ToString()));
-                    }
-
+                    var textBuffer = text.Container.GetTextBuffer();
                     if (textBuffer.Properties.TryGetProperty(typeof(StateMachine), out stateMachine))
                     {
                         TrackingSession trackingSession;
-                        if (stateMachine.CanInvokeRename(out trackingSession, cancellationToken: cancellationToken))
+                        if (!stateMachine.CanInvokeRename(out trackingSession, cancellationToken: cancellationToken))
                         {
-                            var snapshotSpan = stateMachine.TrackingSession.TrackingSpan.GetSpan(stateMachine.Buffer.CurrentSnapshot);
-                            var str = string.Format(EditorFeaturesResources.RenameTo, stateMachine.TrackingSession.OriginalName, snapshotSpan.GetText());
-                            var committerOperation = new RenameTrackingCommitterOperation(new RenameTrackingCommitter(stateMachine, snapshotSpan, _refactorNotifyServices, _undoHistoryRegistry, str, showPreview: _showPreview));
-                            return Task.FromResult(SpecializedCollections.SingletonEnumerable(committerOperation as CodeActionOperation));
+                            Debug.Assert(false, "RenameTracking codefix invoked on a document with a StateMachine which cannot invoke rename.");
+                            return;
                         }
 
-                        Debug.Assert(false, "RenameTracking codefix invoked on a document with a StateMachine which cannot invoke rename.");
+                        var snapshotSpan = stateMachine.TrackingSession.TrackingSpan.GetSpan(stateMachine.Buffer.CurrentSnapshot);
```

Comment:
this seems can have a race. it starts with a snapshot (document) but from there it gets text buffer (mutable) and buffer.CurrentSnapshot which has no guarantee that current snapshot will be same as "document" snapshot.

how is it guarantee session data it gathered is for the "document" snapshot?